lua/lsp-status.lua

```diff
@@ -79,23 +79,22 @@ local function on_attach(client)
   -- Register the client for messages
   messaging.register_client(client.id, client.name)
 
-  -- Set up autocommands to refresh the statusline when information changes
   vim.api.nvim_command('augroup lsp_aucmds')
+  -- Set up autocommands to refresh the statusline when information changes
   vim.api.nvim_command('au! * <buffer>')
-  vim.api.nvim_command('au User LspDiagnosticsChanged redrawstatus!')
-  vim.api.nvim_command('au User LspMessageUpdate redrawstatus!')
-  vim.api.nvim_command('au User LspStatusUpdate redrawstatus!')
-  vim.api.nvim_command('augroup END')
+  vim.api.nvim_command('au User LspDiagnosticsChanged if get(b:, "lsp_status_redraw", 0) | redrawstatus! | end')
+  vim.api.nvim_command('au User LspMessageUpdate if get(b:, "lsp_status_redraw", 0) | redrawstatus! | end')
+  vim.api.nvim_command('au User LspStatusUpdate if get(b:, "lsp_status_redraw", 0) | redrawstatus! | end')
```

Comment:
I used the variable to provide extra checks on the redraw function, like if the status text is the same we shouldn't redraw. Also I didn't know if the `autocmd` was used only by the plugin or was exposed to the user. My reasoning was `event` issued every time there is an update and the redraw only if needed.

Maybe it could be useful to have a parameter in the `update_lsp_statusline(new_status)` that will check if the function need to get the new text from diagnostics and messages with `get_lsp_statusline`.

Also the initial intention was to limit the redraw of the status line every 1/30 seconds since I don't think vim is doing it. I wanted to do it with a check in the `update_lsp_statusline` function that will set the `lsp_status_redraw`. Unfortunately, from my limited understanding of lua, `os.time` function is only in seconds and the  `os.clock` if the cpu time and not the real time, so I didn't add this functionality.

Reply:
I believe this is why coc.nvim doesn't lag, source code link [coc.nvim/src/model/status.ts](https://github.com/neoclide/coc.nvim/blob/0a565835ea666795fa4b17520cd8f148d074d96a/src/model/status.ts#L14)